### PR TITLE
Lazy load chunks

### PR DESCRIPTION
**Pull request by gouthamve**
Wanted to test this first before I pull in all my lazy loading logic.

### CODE REVIEW

**Comment by gouthamve** [comment]
This is good for review, current issues:
1. The chunks are loaded sequentially and one by one and only when accessed. This means when the heap iterator calls `Next()`, the chunks will loaded.
2. We do need to load the chunks for the number of series that exist. That means we will need to load gigabytes into memory.
3. Need to implement `filterChunksByMatchers` again. But first would like a review of this before I do that.

**Comment by gouthamve** [comment]
@tomwilkie This is good for review now!

File: pkg/querier/store.go [hunk 1]
```diff
@@ -92,9 +113,99 @@ outer:
...
+			lastErr = err
+		}
+	}
+
+	return lastErr
+}
+
+func partitionBySeriesChunks(chunks [][]chunk.Chunk, fetchers []*chunk.Fetcher) map[model.Fingerprint][][]chunkenc.LazyChunk {
+	chunksByFp := map[model.Fingerprint][]chunkenc.LazyChunk{}
+	metricByFp := map[model.Fingerprint]model.Metric{}
```
Comment (Kuqd) [inline pkg/querier/store.go]: what is `metricByFp` used for ?

File: pkg/querier/store.go [hunk 2]
```diff
@@ -92,9 +113,99 @@ outer:
...
+	result := make(map[model.Fingerprint][][]chunkenc.LazyChunk, len(chunksByFp))
+
+	for fp, chks := range chunksByFp {
+		result[fp] = partitionOverlappingChunks(chks)
+	}
+
+	return result
+}
+
+func partitionOverlappingChunks(chunks []chunkenc.LazyChunk) [][]chunkenc.LazyChunk {
```
Comment (Kuqd) [inline pkg/querier/store.go]: nitpick: I would like to have a comment, it was hard to follow.

File: pkg/querier/store.go [hunk 3]
```diff
@@ -92,9 +113,99 @@ outer:
...
+
+func loadFirstChunks(ctx context.Context, chks map[model.Fingerprint][][]chunkenc.LazyChunk) error {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "loadFirstChunks")
+	defer sp.Finish()
+
+	// If chunks span buckets, then we'll have different fetchers for each bucket.
+	chksByFetcher := map[*chunk.Fetcher][]*chunkenc.LazyChunk{}
+	for _, lchks := range chks {
+		for _, lchk := range lchks {
+			chksByFetcher[lchk[0].Fetcher] = append(chksByFetcher[lchk[0].Fetcher], &lchk[0])
```
Comment (Kuqd) [inline pkg/querier/store.go]: not sure if possible, but you could have a index out of boundaries.

File: pkg/querier/store.go [hunk 4]
```diff
@@ -92,9 +113,99 @@ outer:
...
+		}
+	}
+
+	sp.LogFields(otlog.Int("fetchers", len(chksByFetcher)))
+
+	errChan := make(chan error)
+	for fetcher, chunks := range chksByFetcher {
+		go func(fetcher *chunk.Fetcher, chunks []*chunkenc.LazyChunk) {
```
Comment (Kuqd) [inline pkg/querier/store.go]: it seems that you expect the first chunk to be available right after this method end as you filter using Metric property from chunk.

However you don't wait for all goroutines to terminate before returning, not sure if this is intended.

Comment (Kuqd) [reply pkg/querier/store.go]: Ho ! you are using the errChan to wait, still you have some code after that sets chunk value that won't be waited for execution.

One way to fix this is to change the code below like this:

```golang
chks, err := fetcher.FetchChunks(ctx, chks, keys)
			
			if err != nil {
                             errChan <- err
				return
			}

 			for i, chk := range chks {
				chunks[i].Chunk = chk
			}
                         errChan <- nil
```

Comment (gouthamve) [reply pkg/querier/store.go]: Nice catch!

File: pkg/querier/store.go (hunk 3 above)
Comment (gouthamve) [reply pkg/querier/store.go]: Fixed it with a check above.

File: pkg/querier/store.go (hunk 2 above)
Comment (gouthamve) [reply pkg/querier/store.go]: Added some comments.
